app/src/test/java/supermarket/ProductTest.kt

```diff
@@ -1,84 +1,303 @@
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts[0].discountAmount, Is(1.00))
     }
 
     @Test
-    fun threeForTwo() {
+    fun `Percent off discount with rounding`() {
+        cart.addItemQuantity(toothbrush, 1.0)
+        teller.addSpecialOffer(PercentageOffer(toothbrush, 10.0))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts[0].discountAmount, Is(0.10))
+    }
+
+    @Test
+    fun `Percent off non-int quantity`() {
+        cart.addItemQuantity(flour, 2.5)
+        teller.addSpecialOffer(PercentageOffer(flour, 50.0))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts[0].discountAmount, Is(3.13))
+    }
+
+    @Test
+    fun `Three for two offer`() {
+        cart.addItemQuantity(apples, 3.0)
+        teller.addSpecialOffer(ThreeForTwoOffer(apples))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts[0].discountAmount, Is(1.99))
+    }
+
+    @Test
+    fun `Three for two with extra product`() {
         cart.addItemQuantity(toothbrush, 4.0)
         teller.addSpecialOffer(ThreeForTwoOffer(toothbrush))
-        val receipt = teller.checksOutArticlesFrom(cart)
-        val discounts = receipt.getDiscounts()
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
         assertThat(discounts.size, Is(1))
-        assertThat(discounts[0].discountAmount, Is(closeTo(0.99, 0.00001)))
-        println(ReceiptPrinter().printReceipt(receipt))
+        assertThat(discounts[0].discountAmount, Is(0.99))
+    }
+
+    @Test
+    fun `Three for two offer without enough product`() {
+        cart.addItemQuantity(apples, 2.0)
+        teller.addSpecialOffer(ThreeForTwoOffer(apples))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts.size, Is(0))
+    }
+
+    @Test
+    fun `Three for two offer applied multiple times`() {
+        cart.addItemQuantity(apples, 8.0)
+        teller.addSpecialOffer(ThreeForTwoOffer(apples))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts.size, Is(1))
+        assertThat(discounts[0].description, containsString("x2"))
+        assertThat(discounts[0].discountAmount, Is(3.98))
+    }
+
+    @Test
+    fun `Three for two offer split over multiple additions`() {
+        cart.addItemQuantity(oranges, 1.0)
+        cart.addItemQuantity(oranges, 2.0)
+        teller.addSpecialOffer(ThreeForTwoOffer(oranges))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts[0].discountAmount, Is(2.00))
+    }
+
+    @Test
```

Comment:
You can just use `@Test(expected = IllegalArgumentException::class)` here for negative testing (hooray for annotations and exceptions!)